### PR TITLE
Add ZueriCrop dataset

### LINKED ISSUE #146: Add ZueriCrop dataset
We should look at adding the [ZueriCrop dataset](https://github.com/0zgur0/ms-convSTAR) from ["Crop mapping from image time series: deep learning with multi-scale label hierarchies", Turkoglu et al. (2021).](https://arxiv.org/abs/2102.08820)

It's a unique dataset for time-series instance segmentation of crops using multispectral Sentinel-2 imagery and 48 fine grained, hierarchical crop categories . It's also quite large in size (roughly 28k time-series of 142 images per time-series, although the images are quite small: 24x24).

### PR DESCRIPTION
**Pull request by isaaccorley**
- Adds `torchgeo.datasets.ZueriCrop` from ["Crop mapping from image time series: deep learning with multi-scale label hierarchies", Turkoglu et al. (2021).](https://arxiv.org/abs/2102.08820)
- Updated docs
- Added hdf5 creation example to `tests/data/README.md`
- Added sample data for unit tests
- Added unit tests

Notes:
- ~~I used `namedtuple` for the hierarchical classes. It looked nice until black formatted them.~~
- Refactor to perform new download/verify check (based on CDL dataset updates)
- Right now I'm outputting a sample dict of image, mask, boxes, and label similar to how [torchvision.models.MaskRCNN](https://pytorch.org/vision/stable/models.html#id41) requires. This is likely the easiest baseline model to use for instance segmentation since `segmentation_models_pytorch` only performs semantic segmentation iirc

Output dimensions (`t` is temporal dim, `n` is number of instances):
- image: (t, c, h, w), t=142 always
- mask: (n, h, w)
- boxes: (n, 4)
- labels: (n, )

Closes #146

### CODE REVIEW

**Comment by adamjstewart** [comment]
Do we need an 827 line list of crop classes? It doesn't seem to be used anywhere in the dataset.

> Right now I'm outputting a sample dict of image, mask, boxes, and label similar to how torchvision.models.MaskRCNN requires. This is likely the easiest baseline model to use for instance segmentation since segmentation_models_pytorch only performs semantic segmentation iirc

If this is an instance segmentation dataset we should use the same format as the NWPU VHR-10 dataset. You can modify that one if the current format isn't ideal.

**Comment by isaaccorley** [comment]
> Do we need an 827 line list of crop classes? It doesn't seem to be used anywhere in the dataset.
> 

No, it's simply a cleaned version of the labels located [here](https://github.com/0zgur0/ms-convSTAR/blob/master/labels.csv). I can download them with the dataset instead.

**Comment by isaaccorley** [comment]
> If this is an instance segmentation dataset we should use the same format as the NWPU VHR-10 dataset. You can modify that one if the current format isn't ideal.

It seems as though the NWPU VHR-10 dataset is only loading the annotation dicts via pycocotools but not generating bboxes or masks. I'll take a look at completing it since pycocotools can actually generate instance masks.

**Comment by adamjstewart** [comment]
> > Do we need an 827 line list of crop classes? It doesn't seem to be used anywhere in the dataset.
> 
> No, it's simply a cleaned version of the labels located [here](https://github.com/0zgur0/ms-convSTAR/blob/master/labels.csv). I can download them with the dataset instead.

Yes, let's do that instead, or ignore the class names entirely if they aren't used in the dataset.

**Comment by isaaccorley** [comment]
I updated the dataset to also download the labels file. It's not used within the dataset, but it's available for users to parse for the hierarchical classes if they want.

File: tests/datasets/test_zuericrop.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import torchgeo.datasets.utils
+from torchgeo.datasets import ZueriCrop
+from torchgeo.transforms import Identity
+
+
+def download_url(url: str, root: str, *args: str, **kwargs: str) -> None:
+    shutil.copy(url, root)
+
+
+class TestZueriCrop:
```
Comment (adamjstewart) [inline tests/datasets/test_zuericrop.py:22]: Should have a `pytest.importorskip` for h5py

File: torchgeo/datasets/zuericrop.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+        self.filepath = os.path.join(root, "ZueriCrop.hdf5")
+
+        self._verify()
+
+        try:
+            import h5py
+        except ImportError:
```
Comment (adamjstewart) [inline torchgeo/datasets/zuericrop.py:89]: Will need tests once you figure things out in your other PR. Let me know if you need help, this was very non-intuitive the first time I did it and there may be a better way.

File: torchgeo/datasets/zuericrop.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            self.size: int = f["data"].shape[0]
+
+    def __getitem__(self, index: int) -> Dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            data and label at that index
```
Comment (adamjstewart) [inline torchgeo/datasets/zuericrop.py:103]:
```suggestion
            sample containing image, mask, bounding boxes, and target label
```

File: torchgeo/datasets/zuericrop.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        self._verify()
+
+        try:
+            import h5py
+        except ImportError:
+            raise ImportError(
+                "h5py is not installed and is required to use this dataset"
+            )
+        with h5py.File(self.filepath, "r") as f:
+            self.size: int = f["data"].shape[0]
```
Comment (adamjstewart) [inline torchgeo/datasets/zuericrop.py:94]: How quick is this? Maybe we should just move it to the `__len__` method. We can cache it if it's slow.

File: torchgeo/datasets/zuericrop.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        """
+        # Check if the files already exist
+        exists = []
+        for filename in self.filenames:
+            filepath = os.path.join(self.root, filename)
+            exists.append(os.path.exists(filepath))
+
+        for e, filename in zip(exists, self.filenames):
+            filepath = os.path.join(self.root, filename)
+            print(e, filepath)
```
Comment (adamjstewart) [inline torchgeo/datasets/zuericrop.py:208]: It looks like this is debugging code that can be removed?

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: torchgeo/datasets/zuericrop.py (hunk 4 above)
Comment (isaaccorley) [reply torchgeo/datasets/zuericrop.py:94]: It's quite fast actually. And `__len__` is only called once per epoch I believe.

File: torchgeo/datasets/zuericrop.py (hunk 5 above)
Comment (isaaccorley) [reply torchgeo/datasets/zuericrop.py]: Removed

File: torchgeo/datasets/zuericrop.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+            RuntimeError: if ``download=False`` but dataset is missing or checksum fails
+        """
+        # Check if the files already exist
+        exists = []
+        for filename in self.filenames:
+            filepath = os.path.join(self.root, filename)
+            exists.append(os.path.exists(filepath))
+
+        for e, filename in zip(exists, self.filenames):
+            filepath = os.path.join(self.root, filename)
```
Comment (adamjstewart) [inline torchgeo/datasets/zuericrop.py:209]: This isn't needed either right?

Comment (isaaccorley) [reply torchgeo/datasets/zuericrop.py:209]: Yup that's correct thanks for catching.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**